src/libs/ReportListUtils.js

```diff
@@ -0,0 +1,313 @@
+});
+
+/**
+ * Get participants for the report. Returns array of logins.
+ *
+ * @param {Object} report
+ * @returns {Array}
+ */
+function getParticipantLogins(report) {
+    return lodashGet(report, ['participants'], []);
+}
+
+/**
+ * Does the report have multiple participants?
+ *
+ * @param {Object} report
+ * @returns {Boolean}
+ */
+function reportHasMultipleParticipants(report) {
+    return getParticipantLogins(report).length > 1;
+}
+
+/**
+ * Is the login provided a report participant?
+ *
+ * @param {Object} report
+ * @param {String} login
+ * @returns {Boolean}
+ */
+function isUserReportParticipant(report, login) {
+    return _.some(getParticipantLogins(report), participantLogin => participantLogin === login);
+}
+
+/**
+ * Check if the report has a draft comment
+ *
+ * @param {Object} comments
+ * @param {Number} reportID
+ * @returns {Boolean}
+ */
+function hasComment(comments, reportID) {
+    const allComments = lodashGet(comments, `${ONYXKEYS.COLLECTION.REPORT_DRAFT_COMMENT}${reportID}`, '');
+    return allComments.length > 0;
+}
+
+/**
+ * Creates a report list option
+ *
+ * @param {Array<Object>} personalDetailList
+ * @param {Object} [report]
+ * @param {Object} [comments]
+ * @param {Number} [activeReportID]
+ * @returns {Object}
+ */
+function createOption(personalDetailList, report, comments, activeReportID) {
+    // There is only one participant so return the single participant view
+    const personalDetail = personalDetailList[0];
+    const isSingleUserDM = personalDetailList.length < 2;
+    return {
+        text: report ? report.reportName : personalDetail.displayName,
+        alternateText: (!report || isSingleUserDM) ? personalDetail.login : report.reportName,
+        icons: report ? report.icons : [personalDetail.avatarURL],
+        login: personalDetailList.length < 2 ? personalDetail.login : null,
+        reportID: report ? report.reportID : null,
+        isUnread: report ? report.unreadActionCount > 0 : null,
+        hasDraftComment: report && report.reportID !== activeReportID && hasComment(comments, report.reportID),
+        keyForList: report ? String(report.reportID) : personalDetail.login,
+    };
+}
+
+function isSearchStringMatch(searchValue, searchText) {
+    const matchRegexes = [
+        new RegExp(`^${Str.escapeForRegExp(searchValue)}$`, 'i'),
+        new RegExp(`^${Str.escapeForRegExp(searchValue)}`, 'i'),
+        new RegExp(Str.escapeForRegExp(searchValue), 'i'),
+    ];
+
+    return _.some(matchRegexes, (regex) => {
+        const valueToSearch = searchText && searchText.replace(new RegExp(/&nbsp;/g), '');
+        return regex.test(valueToSearch);
+    });
+}
+
+/**
+ * Build the options
+ *
+ * @param {Object} reports
+ * @param {Object} personalDetails
+ * @param {Object} reportActions
+ * @param {Object} options
+ * @returns {Object}
+ * @private
+ */
+function getOptions(reports, personalDetails = {}, reportActions = {}, {
```

Comment:
This method is cool but its not what i was envisioning for the future. Hear me out and let me know if this makes sense. So atm in this method is called on each keytype with a new searchValue right, then everytime its called were are checking against a number of logic like `!hasMultipleParticipants && includePersonalDetails && !includeRecentReports` and `!includeRecentReports` and `!includeMultipleParticipantReports && hasMultipleParticipants` etc. But all of those don't need to happen on each keytype and should probably just happen once like when a user was added or removed from a group.

So what I'm thinking is that the moment search is clicked we pass reports and options and then generate allOptions this will then save its values another variable called searchableOptions which saves all the options and then sorts them by pinned and lastVisitedTimestamp using lodashOrderby (we'll have to add the pinned and lastVisitedProperty and maybe even searchText). Now when the user types the first time this method will just do `if (searchValue && !isSearchStringMatch(searchValue, searchText)) {` to check for a match to return or break early if we've got our max results. If a singleUserDM is added and not selected for a group chat then searchableOptions will reject all groupDMReports so that searchableOptions is as light as it can be for the next set of key types.

Same thing goes when a new chat is selected. The only difference is that searchableOptions will not be sorted by pinnedchats (since i see prioritizePinnedReports is false for that flow). 

Does that make sense? Basically trying to keep the actions of keytyping as low as possible. Maybe we can even connect to COLLECTION.REPORT and PERSONAL_DETAILS and generate all options as they change so that when the user clicks on search all the options are already generated and at this point the only thing that changes is sorting logic like sort by pinned, lastVisited etc as needed.

Reply:
Chatted 1:1 to get a better idea of this suggestion. Here's where we landed:

- Subscribe to reports and personalDetails in `ReportSearchUtils`
- Rebuild the options when a change occurs
- Filter when we call one of the public methods (but don't rebuild the options)

Overall it feels nicer. But there's one thing that is bugging me about this now and that is we'll need to also re-render the components when rebuilding options e.g. with the case of the LHN. Not sure whether this will be a problem yet, but something to think about. Converting this to a component could fix that since we can always re-render a child with the options once they are rebuilt. But I won't worry about it now.